File: typescript/src/translator.ts
import { lookupAr, lookupEn, isArabic } from "./lookupIndices";

export function translateToken(token: string, to?: "ar" | "en"): string {
  const srcIsArabic = isArabic(token);
  const target = to || (srcIsArabic ? "en" : "ar");

  if (target === "en") {
    const entry = lookupAr(token);
    return entry ? entry.en : token;
  } else {
    const entry = lookupEn(token);
    return entry ? entry.ar : token;
  }
}

export function translate(fullName: string, to?: "ar" | "en"): string {
  if (!fullName || !fullName.trim()) return fullName;
  const tokens = fullName.trim().split(/\s+/);
  const translated = tokens.map((t) => translateToken(t, to));
  return translated.join(" ");
}

File: java/egy-names/src/main/java/com/afify/egynames/engine/Translator.java
package com.afify.egynames.engine;

import com.afify.egynames.index.LookupIndices;
import com.afify.egynames.model.Models;

import java.util.Arrays;
import java.util.stream.Collectors;

public class Translator {

    public static String translateToken(String token, String to, String dataPath) {
        boolean srcIsAr = LookupIndices.isArabic(token);
        String target = to != null ? to : (srcIsAr ? "en" : "ar");

        if ("en".equalsIgnoreCase(target)) {
            Models.NameEntry entry = LookupIndices.lookupAr(token, dataPath);
            return entry != null ? entry.en : token;
        } else {
            Models.NameEntry entry = LookupIndices.lookupEn(token, dataPath);
            return entry != null ? entry.ar : token;
        }
    }

    public static String translate(String fullName, String to, String dataPath) {
        if (fullName == null || fullName.trim().isEmpty()) return fullName;
        String[] tokens = fullName.trim().split("\\s+");
        return Arrays.stream(tokens)
                .map(t -> translateToken(t, to, dataPath))
                .collect(Collectors.joining(" "));
    }
}
